### PR TITLE
feat(ftintitle): Insert featured artist before track variant clause

### PR DESCRIPTION
**Pull request by treyturner**
## Summary

This PR updates the `ftintitle` plugin to insert featured artist tokens before brackets containing remix/edit-related keywords (e.g., "Remix", "Live", "Edit") instead of always appending them at the end of the title.

## Motivation

Previously, the plugin would always append featured artists at the end of titles, resulting in awkward formatting like:
- `Threshold (Myselor Remix) ft. Hallucinator`

With this change, featured artists are inserted before the first bracket containing keywords, producing cleaner formatting:
- `Threshold ft. Hallucinator (Myselor Remix)`

### Core Functionality

- Added `find_bracket_position()` function that:
  - Searches for brackets containing remix/edit-related keywords
  - Supports multiple bracket types: `()`, `[]`, `<>`, `{}`
  - Only matches brackets with matching opening/closing pairs
  - Uses case-insensitive word-boundary matching for keywords
  - Returns the position of the earliest matching bracket
- Updated `update_metadata()` to insert featured artists before brackets instead of appending

### Configuration

- Added new `bracket_keywords` configuration option:
  - **Default**: List of keywords including: `abridged`, `acapella`, `club`, `demo`, `edit`, `edition`, `extended`, `instrumental`, `live`, `mix`, `radio`, `release`, `remaster`, `remastered`, `remix`, `rmx`, `unabridged`, `unreleased`, `version`, and `vip`
  - **Customizable**: Users can override with their own keyword list
  - **Empty list**: Setting to `[]` matches any bracket content regardless of keywords

### Example Configuration

```yaml
ftintitle:
    bracket_keywords: ["remix", "live", "edit", "version", "extended"]
```

## Behavior

- **Titles with keyword brackets**: Featured artists are inserted before the first bracket containing keywords
  - `Song (Remix) ft. Artist` → `Song ft. Artist (Remix)`
  - `Song (Live) [Remix] ft. Artist` → `Song ft. Artist (Live) [Remix]` (picks first bracket with keyword)

- **Titles without keyword brackets**: Featured artists are appended at the end (backward compatible)
  - `Song (Arbitrary) ft. Artist` → `Song (Arbitrary) ft. Artist`

- **Nested brackets**: Correctly handles nested brackets of same and different types
  - `Song (Remix [Extended]) ft. Artist` → `Song ft. Artist (Remix [Extended])`

- **Multiple brackets**: Picks the earliest bracket containing keywords
  - `Song (Live) (Remix) ft. Artist` → `Song ft. Artist (Live) (Remix)` (if both contain keywords, picks first)

## Testing

- Added comprehensive test coverage for:
  - Different bracket types (`()`, `[]`, `<>`, `{}`)
  - Nested brackets (same and different types)
  - Multiple brackets
  - Custom keywords
  - Empty keyword list behavior
  - Edge cases (unmatched brackets, no brackets, etc.)

All 112 tests pass.

## Backward Compatibility

This change is **backward compatible**:
- Titles without brackets continue to append featured artists at the end
- Titles with brackets that don't contain keywords also append at the end
- Existing configuration files continue to work (uses sensible defaults)

## Documentation

- Updated changelog with detailed description of the new feature
- Configuration option is documented in the changelog entry

### CODE REVIEW

File: beetsplug/ftintitle.py [hunk 1]
```diff
@@ -59,6 +59,90 @@ def contains_feat(title: str, custom_words: list[str] | None = None) -> bool:
...
+    """
+    if keywords is None:
+        keywords = DEFAULT_BRACKET_KEYWORDS
+
+    # If keywords is empty, match any bracket content
+    if not keywords:
+        pattern = None
+    else:
+        # Build regex pattern with word boundaries
+        keyword_pattern = "|".join(rf"\b{re.escape(kw)}\b" for kw in keywords)
```
Comment (snejus) [inline beetsplug/ftintitle.py]: ✍🏼 Nit

```suggestion
        keyword_pattern = rf"\b{'|'.join(map(re.escape, keywords))}\b"
```

File: beetsplug/ftintitle.py [hunk 2]
```diff
@@ -59,6 +59,90 @@ def contains_feat(title: str, custom_words: list[str] | None = None) -> bool:
...
+        pattern = re.compile(keyword_pattern, re.IGNORECASE)
```
[the remark this answers is not in the document]
Comment (treyturner) [reply beetsplug/ftintitle.py]: It does support multi-word phrases, but to your point I added two tests proving it

File: beetsplug/ftintitle.py [hunk 3]
```diff
@@ -59,6 +59,93 @@ def contains_feat(title: str, custom_words: list[str] | None = None) -> bool:
...
+
+    # Bracket pairs (opening, closing)
+    bracket_pairs = [("(", ")"), ("[", "]"), ("<", ">"), ("{", "}")]
+
+    # Track the earliest valid bracket position
+    earliest_pos = None
+
+    for open_char, close_char in bracket_pairs:
```
Comment (snejus) [inline beetsplug/ftintitle.py]: It seems to me this could be achieved with a pattern as well?

```py
    # If keywords is empty, match any bracket content
    if not keywords:
        keyword_pat = ".*?"
    else:
        # Build regex pattern to support multi-word keywords/phrases.
        # Each keyword/phrase is escaped and surrounded by word boundaries at
        # start and end, matching phrases like "club mix" as a whole.
        keyword_pat = "|".join(rf"\b{re.escape(kw)}\b" for kw in keywords)

    pattern = re.compile(
        rf"""
        \(.*?({keyword_pat}).*?\) |
        \[.*?({keyword_pat}).*?\] |
        <.*?({keyword_pat}).*?> |
        \{{.*?({keyword_pat}).*?\}}
        """,
        re.IGNORECASE | re.VERBOSE,
    )

    return m.start() if (m := pattern.search(title)) else None
```

File: beetsplug/ftintitle.py [hunk 4]
```diff
@@ -59,6 +59,93 @@ def contains_feat(title: str, custom_words: list[str] | None = None) -> bool:
...
+    """Find the position of the first opening bracket that contains
+    remix/edit-related keywords and has a matching closing bracket.
+
+    Args:
+        title: The title to search in.
+        keywords: List of keywords to match. If None, uses DEFAULT_BRACKET_KEYWORDS.
+            If an empty list, matches any bracket content (not just keywords).
+
+    Returns:
+        The position of the opening bracket, or None if no match found.
```
Comment (snejus) [inline beetsplug/ftintitle.py]: You've made parameter names and types in the function signature clear enough, so this is redundant and vulnerable to getting out of sync.

```suggestion
```

File: beetsplug/ftintitle.py [hunk 5]
```diff
@@ -138,6 +226,7 @@ def func(lib, opts, args):
                 bool
             )
             custom_words = self.config["custom_words"].get(list)
+            bracket_keywords = self.config["bracket_keywords"].get(list)
```
Comment (snejus) [inline beetsplug/ftintitle.py]:
```suggestion
            bracket_keywords = self.config["bracket_keywords"].as_str_seq()
```

File: beetsplug/ftintitle.py [hunk 6]
```diff
@@ -208,7 +301,14 @@ def update_metadata(
         if not drop_feat and not contains_feat(item.title, custom_words):
             feat_format = self.config["format"].as_str()
             new_format = feat_format.format(feat_part)
-            new_title = f"{item.title} {new_format}"
+            # Insert before the first bracket containing remix/edit keywords
+            bracket_pos = find_bracket_position(item.title, bracket_keywords)
```
Comment (snejus) [inline beetsplug/ftintitle.py]: I'd suggest defining `find_bracket_position` as a plugin method to make it responsible for resolution of `bracket_keywords`

```suggestion
            bracket_pos = self.find_bracket_position(item.title)
```

This way, you don't need to send `bracket_keywords` in intermediate method calls.

File: beetsplug/ftintitle.py [hunk 7]
```diff
@@ -208,7 +301,14 @@ def update_metadata(
...
+            if bracket_pos is not None:
+                title_before = item.title[:bracket_pos].rstrip()
+                title_after = item.title[bracket_pos:]
+                new_title = f"{title_before} {new_format} {title_after}"
+            else:
+                new_title = f"{item.title} {new_format}"
```
Comment (snejus) [inline beetsplug/ftintitle.py]: It's probably a good idea to define a new method to have this entire logic in a single place

```suggestion
            new_title = self.insert_ft_into_title(title, new_format)
```

```py
...

    def insert_ft_into_title(self, title: str, ft_part:  str) -> str:
        """Make sure the ft part is inserted before any brackets in the title."""
        ...
```

You may as well merge this with `find_bracket_position` logic.

File: beetsplug/ftintitle.py [hunk 8]
```diff
@@ -161,6 +251,7 @@ def imported(self, session: ImportSession, task: ImportTask) -> None:
         keep_in_artist_field = self.config["keep_in_artist"].get(bool)
         preserve_album_artist = self.config["preserve_album_artist"].get(bool)
         custom_words = self.config["custom_words"].get(list)
+        bracket_keywords = self.config["bracket_keywords"].get(list)
```
Comment (snejus) [inline beetsplug/ftintitle.py]: You can DRY this up by defining a cached property to resolve them:

```py
    @cached_property
    def bracket_keywords(self) -> list[str] | None:
        return self.config["bracket_keywords"].as_str_seq()
```

**Review by snejus — CHANGES_REQUESTED** [review]
Added a couple of suggestions to slightly simplify the structure

**Comment by treyturner** [comment]
> Added a couple of suggestions to slightly simplify the structure

Thanks, will get to these when I can!

File: beetsplug/ftintitle.py (hunk 3 above)
Comment (treyturner) [reply beetsplug/ftintitle.py]: Yes, very much how this should have originally been written 👍 Thank you for cleaning it up.

**Comment by treyturner** [comment]
@snejus Thanks for the review, these were all good points. I took a stab at addressing them but am happy to respond to anything else you find.

File: beetsplug/ftintitle.py [hunk 9]
```diff
@@ -99,6 +125,10 @@ def find_feat_part(
 
 
 class FtInTitlePlugin(plugins.BeetsPlugin):
+    @cached_property
```
Comment (snejus) [inline beetsplug/ftintitle.py:132]: Much cleaner!

File: beetsplug/ftintitle.py [hunk 10]
```diff
@@ -250,6 +281,52 @@ def ft_in_title(
...
-            item, feat_part, drop_feat, keep_in_artist_field, custom_words
+            item,
+            feat_part,
+            drop_feat,
+            keep_in_artist_field,
+            custom_words,
         )
         return True
+
+    def find_bracket_position(
```
Comment (snejus) [inline beetsplug/ftintitle.py:329]: What would you think about merging these two methods? `find_bracket_position` has no use outside this context.

File: beetsplug/ftintitle.py [hunk 11]
```diff
@@ -250,6 +281,52 @@ def ft_in_title(
...
+        keywords = self.bracket_keywords
+
+        # If keywords is empty, match any bracket content
+        if not keywords:
+            keyword_ptn = ".*?"
+        else:
+            # Build regex supporting keywords/multi-word phrases.
+            keyword_ptn = rf"\b{'|'.join(map(re.escape, keywords))}\b"
+
+        pattern = re.compile(
```
Comment (snejus) [inline beetsplug/ftintitle.py]: I'm looking at this, and it seems that all conditional logic could be delegated to `re` 😆 

```py
        pattern = re.compile(
            rf"""
            (?:   # non-capturing group for the split
              \s*?  # optional whitespace before brackets
              (?=     # any bracket containing a keyword
                \(.*?{keyword_ptn}.*?\) |
                \[.*?{keyword_ptn}.*?\] |
                 <.*?{keyword_ptn}.*? > |
                {{.*?{keyword_ptn}.*?}} |
                $     # or the end of the string
              )
            )
            """,
            re.IGNORECASE | re.VERBOSE,
        )
        parts = pattern.split(title, maxsplit=1)
        return f" {feat_part} ".join(parts).strip()
```

I like the idea of removing `if ... else` from here, but this requires a more advanced understanding of regular expressions.

Which version would you prefer seeing if you came across this randomly as an external contributor?

File: beetsplug/ftintitle.py [hunk 12]
```diff
@@ -110,6 +140,7 @@ def __init__(self) -> None:
                 "keep_in_artist": False,
                 "preserve_album_artist": True,
                 "custom_words": [],
+                "bracket_keywords": DEFAULT_BRACKET_KEYWORDS.copy(),
```
Comment (snejus) [inline beetsplug/ftintitle.py]: Simply define them inline instead of using the constant - they aren't (or shouldn't be) used anywhere else.
```suggestion
                "bracket_keywords": [
                    ...
                ],
```

File: beetsplug/ftintitle.py [hunk 13]
```diff
@@ -250,6 +281,52 @@ def ft_in_title(
 
         # If we have a featuring artist, move it to the title.
         self.update_metadata(
-            item, feat_part, drop_feat, keep_in_artist_field, custom_words
+            item,
+            feat_part,
+            drop_feat,
```
Comment (snejus) [inline beetsplug/ftintitle.py]: Seems to be unrelated to your adjustment?

File: test/plugins/test_ftintitle.py [hunk 14]
```diff
@@ -246,6 +270,64 @@ def add_item(
...
+            ("Alice", "Song 1 ft. Bob (Club Mix)"),
+            id="multi-word-keyword-positive-match",
+        ),
+        pytest.param(
+            {"format": "ft. {}", "bracket_keywords": ["club mix"]},
+            ("ftintitle",),
+            ("Alice ft. Bob", "Song 1 (Club Remix)", "Alice"),
+            ("Alice", "Song 1 (Club Remix) ft. Bob"),
+            id="multi-word-keyword-negative-no-match",
+        ),
```
Comment (snejus) [inline test/plugins/test_ftintitle.py:266]: There is no need to test different variations in a functional test - that is already well covered by unit tests.

You can leave a pair of tests cases to _precisely_ display what does this configuration do. I'd specifically want to see these two cases:

```suggestion
        pytest.param(
            {"format": "ft. {}", "bracket_keywords": ["mix"]},
            ("ftintitle",),
            ("Alice ft. Bob", "Song 1 (Club Mix)", "Alice"),
            ("Alice", "Song 1 ft. Bob (Club Mix)"),
            id="ft-inserted-before-matching-bracket-keyword",
        ),
        pytest.param(
            {"format": "ft. {}", "bracket_keywords": ["nomatch"]},
            ("ftintitle",),
            ("Alice ft. Bob", "Song 1 (Club Remix)", "Alice"),
            ("Alice", "Song 1 (Club Remix) ft. Bob"),
            id="ft-inserted-at-end-no-bracket-keyword-match",
        ),

```

File: test/plugins/test_ftintitle.py [hunk 15]
```diff
@@ -269,6 +351,31 @@ def test_ftintitle_functional(
     assert item["title"] == expected_title
 
 
+def test_imported_stage_moves_featured_artist(
```
Comment (snejus) [inline test/plugins/test_ftintitle.py]: I doubt this is within the scope of your adjustment. We already have multiple ways of testing imports across the test codebase, so we should reuse those methods. In any case, I think it's best to leave this for another PR.

File: test/plugins/test_ftintitle.py [hunk 16]
```diff
@@ -312,6 +419,83 @@ def test_split_on_feat(
...
+        ("Song (Anything)", [], 5),
+        ("Song (Remix)", [], 5),
+        ("Song", [], None),
+        ("Song (", [], None),
+        # Multi-word keyword tests
+        ("Song (Club Mix)", ["club mix"], 5),  # Positive: matches multi-word
+        ("Song (Club Remix)", ["club mix"], None),  # Negative: no match
+    ],
+)
+def test_find_bracket_position_custom_keywords(
```
Comment (snejus) [inline test/plugins/test_ftintitle.py]: Let's join these two tests into a single one: `test_insert_ft_into_title`.

File: test/plugins/test_ftintitle.py [hunk 17]
```diff
@@ -39,19 +43,29 @@ def env() -> Generator[FtInTitlePluginFunctional, None, None]:
...
         "auto": True,
         "keep_in_artist": False,
         "custom_words": [],
+        "bracket_keywords": ftintitle.DEFAULT_BRACKET_KEYWORDS.copy(),
     }
     env.config["ftintitle"].set(defaults)
     env.config["ftintitle"].set(cfg)
 
 
+def build_plugin(
```
Comment (snejus) [inline test/plugins/test_ftintitle.py]: Your unit tests should ideally not depend on the plugin being initialised. That's what functional tests are for.

I think, a good way around this is to define `insert_ft_into_title` as a static method:
```py
    @staticmethod
    def insert_ft_into_title(
        title: str, feat_part: str, keywords: list[str]
    ) -> str:
```

And now you can test it without having to depend on plugin init or needing to manage the config:

```py
assert FtInTitlePlugin.insert_ft_into_title(title, ft_part, keywords) ...
```

File: test/plugins/test_ftintitle.py [hunk 18]
```diff
@@ -39,19 +43,29 @@ def env() -> Generator[FtInTitlePluginFunctional, None, None]:
...
-    cfg: dict[str, str | bool | list[str]] | None,
+    cfg: dict[str, ConfigValue] | None,
 ) -> None:
     cfg = {} if cfg is None else cfg
     defaults = {
         "drop": False,
         "auto": True,
         "keep_in_artist": False,
         "custom_words": [],
+        "bracket_keywords": ftintitle.DEFAULT_BRACKET_KEYWORDS.copy(),
```
Comment (snejus) [inline test/plugins/test_ftintitle.py]: I think the default config is already set by the plugin?

**Review by snejus — CHANGES_REQUESTED** [review]
Thank you for your adjustments! Another round of comments - I've covered everything now :)

File: beetsplug/ftintitle.py (hunk 13 above)
Comment (treyturner) [reply beetsplug/ftintitle.py]: A formatter wanted this, but it may have been triggered when the additional property exceeded line-length. Will see if it allows me to revert.

File: beetsplug/ftintitle.py (hunk 10 above)
Comment (treyturner) [reply beetsplug/ftintitle.py:329]: Can't say I've dug through enough of the whole codebase for my feedback to be very meaningful. It was an intuitive concern to separate and I gambled that letting contributors choose which to dabble in would be nice. 🤷‍♂️ I'm also thinking I or someone else may look to expand cases and/or improve handling in `insert_ft_into_title()` soon.

Happy to combine them until that actually happens though if this feels overengineered.

File: test/plugins/test_ftintitle.py (hunk 15 above)
Comment (treyturner) [reply test/plugins/test_ftintitle.py]: Can't blame me for trying to cover the one uncovered line, but understood, will remove ;)

File: test/plugins/test_ftintitle.py (hunk 17 above)
Comment (treyturner) [reply test/plugins/test_ftintitle.py]: Good lookin' out, thanks!

File: beetsplug/ftintitle.py (hunk 10 above)
Comment (treyturner) [reply beetsplug/ftintitle.py:329]: I ended up leaving it separate as it's more testable this way.

File: test/plugins/test_ftintitle.py (hunk 16 above)
Comment (treyturner) [reply test/plugins/test_ftintitle.py]: Done!

File: beetsplug/ftintitle.py (hunk 11 above)
Comment (treyturner) [reply beetsplug/ftintitle.py]: I went with a slightly different pattern and opted to cache it since it's a per-track operation. Let me know if you find any issues with the new approach, and thanks for your patience and help!

File: beetsplug/ftintitle.py (hunk 12 above)
Comment (treyturner) [reply beetsplug/ftintitle.py]: I removed this per your request initially, but ended up restoring it so the test could access the defaults in a static context 😅

**Comment by treyturner** [comment]
Ok, ready for another round I think @snejus. Sorry for the delay! The longer it took me to get back to this, the longer it was taking to resolve the comments.

File: test/plugins/test_ftintitle.py [hunk 19]
```diff
@@ -312,6 +330,64 @@ def test_split_on_feat(
     assert ftintitle.split_on_feat(given) == expected
 
 
+@pytest.mark.parametrize(
```
Comment (snejus) [inline test/plugins/test_ftintitle.py:333]: I think we'd prefer testing the outermost function

```py
@pytest.mark.parametrize(
    "given,keywords,expected",
    [
        ## default keywords
        # different braces and keywords
        ("Song (Remix)", None, "Song ft. Bob (Remix)"),
        ("Song [Version]", None, "Song ft. Bob [Version]"),
        ("Song {Extended Mix}", None, "Song ft. Bob {Extended Mix}"),
        ("Song <Instrumental>", None, "Song ft. Bob <Instrumental>"),
        # two keyword clauses
        ("Song (Remix) (Live)", None, "Song ft. Bob (Remix) (Live)"),
        # brace insensitivity
        ("Song (Live) [Remix]", None, "Song ft. Bob (Live) [Remix]"),
        ("Song [Edit] (Remastered)", None, "Song ft. Bob [Edit] (Remastered)"),
        # negative cases
        ("Song", None, "Song ft. Bob"),  # no clause
        ("Song (Arbitrary)", None, "Song (Arbitrary) ft. Bob"),  # no keyword
        ("Song (", None, "Song ( ft. Bob"),  # no matching brace or keyword
        ("Song (Live", None, "Song (Live ft. Bob"),  # no matching brace with keyword
        # one keyword clause, one non-keyword clause
        ("Song (Live) (Arbitrary)", None, "Song ft. Bob (Live) (Arbitrary)"),
        ("Song (Arbitrary) (Remix)", None, "Song (Arbitrary) ft. Bob (Remix)"),
        # nested brackets - same type
        ("Song (Remix (Extended))", None, "Song ft. Bob (Remix (Extended))"),
        ("Song [Arbitrary [Description]]", None, "Song [Arbitrary [Description]] ft. Bob"),
        # nested brackets - different types
        ("Song (Remix [Extended])", None, "Song ft. Bob (Remix [Extended])"),
        # nested - returns outer start position despite inner keyword
        ("Song [Arbitrary {Extended}]", None, "Song ft. Bob [Arbitrary {Extended}]"),
        ("Song {Live <Arbitrary>}", None, "Song ft. Bob {Live <Arbitrary>}"),
        ("Song <Remaster (Arbitrary)>", None, "Song ft. Bob <Remaster (Arbitrary)>"),
        ("Song <Extended> [Live]", None, "Song ft. Bob <Extended> [Live]"),
        ("Song (Version) <Live>", None, "Song ft. Bob (Version) <Live>"),
        ("Song (Arbitrary [Description])", None, "Song (Arbitrary [Description]) ft. Bob"),
        ("Song [Description (Arbitrary)]", None, "Song [Description (Arbitrary)] ft. Bob"),
        ## custom keywords
        ("Song (Live)", ["live"], "Song ft. Bob (Live)"),
        ("Song (Concert)", ["concert"], "Song ft. Bob (Concert)"),
        ("Song (Remix)", ["custom"], "Song (Remix) ft. Bob"),
        ("Song (Custom)", ["custom"], "Song ft. Bob (Custom)"),
        ("Song", [], "Song ft. Bob"),
        ("Song (", [], "Song ( ft. Bob"),
        # Multi-word keyword tests
        ("Song (Club Mix)", ["club mix"], "Song ft. Bob (Club Mix)"),  # Positive: matches multi-word
        ("Song (Club Remix)", ["club mix"], "Song (Club Remix) ft. Bob"),  # Negative: no match
    ],
)  # fmt: skip
def test_insert_ft_into_title(
    given: str,
    keywords: list[str] | None,
    expected: str,
) -> None:
    assert (
        ftintitle.FtInTitlePlugin.insert_ft_into_title(
            given, "ft. Bob", keywords
        )
        == expected
    )
```

1. This tests the _behaviour_ rather than implementation details
2. The expected behaviour is much more clear to maintainers.

Note the presence of `# fmt: skip` - we want to keep each test case in a single line for clarity. Just add this to `pyproject.toml`:

```toml
[tool.ruff.lint.per-file-ignores]
"beets/**" = ["PT"]
"test/test_util.py" = ["E501"]
"test/plugins/test_ftintitle.py" = ["E501"]
```

File: beetsplug/ftintitle.py [hunk 20]
```diff
@@ -104,6 +129,53 @@ def _album_artist_no_feat(album: Album) -> str:
...
+              >                     # ">"
+
+            | \x7B                  # Literal open brace
+                (?=[^\x7D]*{kw})    # Lookahead
+                [^\x7D]*            # Consume content up to first close brace
+              \x7D                  # Literal close brace
+            )                       # End bracketed segment alternation
+            """,
+            re.IGNORECASE | re.VERBOSE,
+        )
```
Comment (snejus) [inline beetsplug/ftintitle.py:164]: This has got a bit verbose, can't we just have the following instead?

```suggestion
        return re.compile(
            rf"""
            \([^)]*({kw}).*?\) |
            \[[^]]*({kw}).*?\] |
            <[^>]*({kw}).*?> |
            \{{[^}}]*({kw}).*?\}}
            """,
            re.IGNORECASE | re.VERBOSE,
        )
```

Comment (snejus) [inline beetsplug/ftintitle.py]: This has got a bit verbose, can't we have the following instead?

```suggestion
        kw = rf"\b(?={kw_inner})\b" if kw_inner else ""
        return re.compile(
            rf"""
            (?:   # non-capturing group for the split
              \s*?  # optional whitespace before brackets
              (?=     # any bracket containing a keyword
                    \([^)]*{kw}.*?\)
                |   \[[^]]*{kw}.*?\]
                |    <[^>]*{kw}.*? >
                | \{{[^}}]*{kw}.*?\}}
                | $   # or the end of the string
              )
            )
            """,
            re.IGNORECASE | re.VERBOSE,
        )
```

**Note**: I updated my initial suggestion to make sure it splits properly:

```py
>>> FtInTitlePlugin._bracket_position_pattern(("live",)).split("hello (live)", maxsplit=1)
['hello', '(live)']

>>> FtInTitlePlugin._bracket_position_pattern(("live",)).split("hello", maxsplit=1)
['hello', '']

>>> " ft. Artist ".join(FtInTitlePlugin._bracket_position_pattern(("live",)).split("hello (live)", maxsplit=1))
'hello ft. Artist (live)'

>>> " ft. Artist ".join(FtInTitlePlugin._bracket_position_pattern(("live",)).split("hello", maxsplit=1)).strip()
'hello ft. Artist'

>>> " ft. Artist ".join(FtInTitlePlugin._bracket_position_pattern(("live",)).split("hello (live) [remix]", maxsplit=1)).strip()
'hello ft. Artist (live) [remix]'

>>> " ft. Artist ".join(FtInTitlePlugin._bracket_position_pattern(("live",)).split("hello [remix] (live)", maxsplit=1)).strip()
'hello [remix] ft. Artist (live)'

>>> " ft. Artist ".join(FtInTitlePlugin._bracket_position_pattern(("live",)).split("hello [remix (live)", maxsplit=1)).strip()
'hello [remix ft. Artist (live)'
```

File: beetsplug/ftintitle.py [hunk 21]
```diff
@@ -216,8 +289,10 @@ def update_metadata(
         # artist and if we do not drop featuring information.
         if not drop_feat and not contains_feat(item.title, custom_words):
             feat_format = self.config["format"].as_str()
-            new_format = feat_format.format(feat_part)
-            new_title = f"{item.title} {new_format}"
+            formatted = feat_format.format(feat_part)
+            new_title = FtInTitlePlugin.insert_ft_into_title(
```
Comment (snejus) [inline beetsplug/ftintitle.py]:
```suggestion
            new_title = self.insert_ft_into_title(
```

File: beetsplug/ftintitle.py (hunk 10 above)
Comment (snejus) [reply beetsplug/ftintitle.py:329]: Added a suggestion regarding the test. It allows merging the methods:

```py
    @classmethod
    def insert_ft_into_title(
        cls, title: str, feat_part: str, keywords: list[str] | None = None
    ) -> str:
        """Insert featured artist before the first bracket containing
        remix/edit keywords if present.
        """
        normalized = (
            DEFAULT_BRACKET_KEYWORDS if keywords is None else tuple(keywords)
        )
        pattern = cls._bracket_position_pattern(normalized)
        parts = pattern.split(title, maxsplit=1)
        return f" {feat_part} ".join(parts).strip()
```

Tested that it works locally. Note, since it depends on another method defined on this class, it makes sense to make it a `@classmethod` so that you can use `cls` instead of hard-coding the class name.

**Review by snejus — CHANGES_REQUESTED** [review]
Added a couple of suggestions.

Apologies for being so needy 😅 I just want to make sure that the code you will merge in won't ever need to get adjusted :)

**Comment by treyturner** [comment]
Oops, didn't mean to ping again if that ended up happening... Just couldn't tell if I had marked the change as (hopefully) addressed.

**Review by snejus — APPROVED** [review]
This looks great, thanks for your patience @treyturner! <3
